src/twisted/internet/test/test_inlinecb.py

```diff
@@ -3,22 +3,332 @@
 # See LICENSE for details.
 
 """
-Tests for L{twisted.internet.defer.inlineCallbacks}.
-
-Some tests for inlineCallbacks are defined in L{twisted.test.test_defgen} as
-well.
+Tests for L{twisted.internet.inlineCallbacks}.
 """
 
+import traceback
+import unittest as pyunit
+import weakref
+from typing import Any, Generator, List, Set, Union
 
+from twisted.internet import reactor, task
 from twisted.internet.defer import (
     CancelledError,
     Deferred,
+    Failure,
+    fail,
     inlineCallbacks,
     returnValue,
+    succeed,
 )
+from twisted.python.compat import _PYPY
 from twisted.trial.unittest import SynchronousTestCase, TestCase
 
 
+def getThing():
+    d = Deferred()
+    reactor.callLater(0, d.callback, "hi")
+    return d
+
+
+def getDivisionFailure(msg: Union[str, None] = None) -> Failure:
+    """
+    Make a L{Failure} of a divide-by-zero error.
+    """
+    try:
+        raise ZeroDivisionError(msg)
+    except BaseException:
+        f = Failure()
+    return f
+
+
+class TerminalException(Exception):
+    pass
+
+
+class BasicTests(TestCase):
+    # First provide all the generator impls necessary for BaseDefgenTests
+
+    @inlineCallbacks
+    def _genBasics(self):
+        x = yield getThing()
+
+        self.assertEqual(x, "hi")
+
+        try:
+            yield getDivisionFailure("OMG")
+        except ZeroDivisionError as e:
+            self.assertEqual(str(e), "OMG")
+        returnValue("WOOSH")
+
+    def testBasics(self):
+        """
+        Test that a normal inlineCallbacks works.  Tests yielding a
+        deferred which callbacks, as well as a deferred errbacks. Also
+        ensures returning a final value works.
+        """
+
+        return self._genBasics().addCallback(self.assertEqual, "WOOSH")
+
+    @inlineCallbacks
+    def _genBuggy(self):
+        yield getThing()
+        1 / 0
+
+    def testBuggy(self):
+        """
+        Ensure that a buggy generator properly signals a Failure
+        condition on result deferred.
+        """
+        return self.assertFailure(self._genBuggy(), ZeroDivisionError)
+
+    @inlineCallbacks
+    def _genNothing(self):
+        if False:
+            yield 1
+
+    def testNothing(self):
+        """Test that a generator which never yields results in None."""
+
+        return self._genNothing().addCallback(self.assertEqual, None)
```

Comment:
Do we need to spin the reactor for these tests ?

This comment is valid for the other tests.

If spining the reactor is needed, I think that is important to add a comment.


Maybe something like this?

```suggestion
        """
        C{None} is returned by a generator that never yields.
        """
        deferred = self._genNothing()
        result = self.successResultOf(deferred)
        self.assertIsNone(result)

```

Reply:
It is important to directly exercise the API because inlineCallbacks and Deferred may be optimized to interact in specific ways depending on how they are invoked together. Therefore it is important to specifically use addCallback here.

It would be a more complete test if there was a sibling of this test that tested the case of something in _genNothing blocking.

Reply:
I added a note explaining this.